File: Cpp/1805_Number_of_Different_Integers_in_a_String.cpp
class Solution {
public:
    int numDifferentIntegers(string word) {
        unordered_set<string> s;
        string num = "";

        for(int i = 0; i < word.size(); ++i) {
            if(word[i] == '0' && num.empty() && i < word.size() - 1 && isdigit(word[i + 1])) continue;

            if(isdigit(word[i])) {
                num += word[i];
            } else if(i > 0 && isdigit(word[i - 1])) {
                s.insert(num);
                num = "";
            }
        }

        if(isdigit(word.back())) s.insert(num);

        return s.size();
    }
};

File: Kotlin/1805_Number_of_Different_Integers_in_a_String.kt
class Solution {
    fun numDifferentIntegers(word: String): Int {
        val s = HashSet<String>()
        var num = ""
        
        for(i in word.indices) {
            if(word[i] == '0' && num.isEmpty() && i < word.length - 1 && word[i + 1].isDigit()) continue

            if(word[i].isDigit()) {
                num += word[i]
            } else if(i > 0 && word[i - 1].isDigit()) {
                s.add(num)
                num = ""
            }
        }
        
        if(word.last().isDigit()) s.add(num)

        return s.size
    }
}